DOC Adds Release Highlights for 1.4

thomasjpfan opened the pull request:
This PR starts off the release notes for 1.4.

thomasjpfan committed:
DOC Adds initial release notes for 1.4

adrinjalali committed:
add metadata routing example

adrinjalali commented:
Added a metadata routing example, feel free to edit.

glemaitre commented on examples/release_highlights/plot_release_highlights_1_4_0.py:
```diff
@@ -0,0 +1,103 @@
+from sklearn.model_selection import GridSearchCV, cross_validate, GroupKFold
+import numpy as np
+
+sklearn.set_config(enable_metadata_routing=True)
```
You might want to disable at the end of the portion of code.

adrinjalali replied:
added with a comment

adrinjalali committed:
add comment and disable metadata-routing flag at the end of script

glemaitre commented:
Here is my wish list in order of module appearance:

- Polars support
- HTML representation
- Ongoing work in Array API (if we decide that there is not enough support we can wait to communicate on it)
- PCA on sparse container
- Missing values for tree-based model (new in ensemble)
- Monotonic constraint in all tree-based models

adrinjalali commented:
cc @scikit-learn/core-devs 

It's easier if each of us adds the section for which we've been working on.

jjerphan commented:
> PCA on sparse container

I can participate or do this one.

betatim commented:
I'm +-0 for adding Array API things to the release notes. What do others think?

thomasjpfan committed:
Adds missing value support for random forest entry

thomasjpfan committed:
Merge remote-tracking branch 'upstream/main' into release_highlights

thomasjpfan committed:
Add polars item

thomasjpfan committed:
Merge remote-tracking branch 'upstream/main' into release_highlights

thomasjpfan committed:
Update env

thomasjpfan committed:
Use min polars in min doc build

thomasjpfan commented:
I updated the PR and added an entry for missing values support in Random Forest and Polars.

For the polars entry, I had to update the lock files to include polars in the doc building.

GaelVaroquaux commented on examples/release_highlights/plot_release_highlights_1_4_0.py:
```diff
@@ -0,0 +1,149 @@
+
+# %%
+# Metadata Routing Support
+# ------------------------
```
I think that this should be further down the line: it's quite technical

GaelVaroquaux commented on examples/release_highlights/plot_release_highlights_1_4_0.py:
```diff
@@ -0,0 +1,149 @@
+hist.fit(X_train, y_train)
+y_decision = hist.decision_function(X_test)
+print(f"ROC AUC score is {roc_auc_score(y_test, y_decision)}")
+
```
```suggestion

# %%
# Enriched estimator displays
# --------------------------------------
# Estimators displays have been enriched: if we look at `hist`, defined above:
hist

# %%
# There is a link to the document in the upper right of the estimator.
# In addition, the display changes color, from orange to blue, when the estimator is fitted.
```

GaelVaroquaux reviewed (commented):
Add estimator displays

glemaitre committed:
Merge branch 'main' into release_highlights

glemaitre committed:
Update examples/release_highlights/plot_release_highlights_1_4_0.py

Co-authored-by: Gael Varoquaux <gael.varoquaux@normalesup.org>

glemaitre committed:
reorder

glemaitre committed:
add monotonic constraint

glemaitre committed:
force plotting

glemaitre commented:
I added a section on the monotonic constraint in tree-based models. I just retook the example from a previous release.

adrinjalali reviewed (approved):
LGTM. Should we merge and have it in the RC and add before the release?

glemaitre commented:
> LGTM. Should we merge and have it in the RC and add before the release?

I would be inclined with this solution. It would give @jjerphan some time to propose a paragraph for the sparse PCA speed-up.

jjerphan commented:
This works for me. I should be able to do that in 2 weeks.

lorentzenchr commented:
Should we mention `max_features` in HGBT?

jeremiedbb commented on examples/release_highlights/plot_release_highlights_1_4_0.py:
```diff
@@ -0,0 +1,206 @@
+# %%
+# Enriched estimator displays
+# ---------------------------
+# Estimators displays have been enriched: if we look at `hist`, defined above:
```
```suggestion
# Estimators displays have been enriched: if we look at `forest`, defined above:
```

jeremiedbb reviewed (approved):
LGTM. Let's merge as is as discussed and we'll add the PCA section (and possibly more) in between the RC and the final.

jeremiedbb committed:
Update examples/release_highlights/plot_release_highlights_1_4_0.py

jjerphan commented:
> This works for me. I should be able to do that in 2 weeks.

I started adding some notes in https://github.com/scikit-learn/scikit-learn/pull/28138. I am not sure when I will be able to focus on writing the rest unfortunately.
